spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoDateFactory.scala

```diff
@@ -0,0 +1,88 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.conformance.streaming
+
+import org.apache.commons.configuration2.Configuration
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{DateType, TimestampType, StringType}
+import org.apache.spark.sql.{Column, DataFrame}
+import org.slf4j.{Logger, LoggerFactory}
+import za.co.absa.enceladus.common.Constants.ReportDateFormat
+import za.co.absa.enceladus.conformance.HyperConformance.{eventTimestampColumnKey, eventTimestampPatternKey, reportDateKey}
+import za.co.absa.enceladus.conformance.datasource.PartitioningUtils
+import za.co.absa.enceladus.utils.schema.SchemaUtils
+
+/**
+ * Info date factory allows to create an expression for the information date based on the strategy used.
+ * It is one of:
+ * - Have a fixed information date provided by the configuration (or Hyperdrive command line)
+ * - Derive it from an event timestamp column
+ * - Use the processing time
+ */
+sealed trait InfoDateFactory {
+  def getInfoDateColumn(df: DataFrame): Column
+}
+
+class InfoDateLiteralFactory(infoDate: String) extends InfoDateFactory {
+  override def getInfoDateColumn(df: DataFrame): Column = {
+    PartitioningUtils.validateReportDate(infoDate)
+    to_date(lit(infoDate), ReportDateFormat)
+  }
+}
+
+class InfoDateFromColumnFactory(columnName: String, pattern: String) extends InfoDateFactory {
+  override def getInfoDateColumn(df: DataFrame): Column = {
+    val dt = SchemaUtils.getFieldType(columnName, df.schema)
+    dt match {
+      case Some(TimestampType) =>
+        col(columnName).cast(DateType)
+      case Some(DateType) =>
+        col(columnName)
+      case Some(StringType) =>
+        to_timestamp(col(columnName), pattern).cast(DateType)
+      case Some(_) =>
+        throw new IllegalArgumentException(s"The specified event time column $columnName has an incompatible type: $dt")
+      case None =>
+        throw new IllegalArgumentException(s"The specified event time column does not exist: $columnName")
+    }
+  }
+}
+
+class InfoDateFromProcessingTimeFactory extends InfoDateFactory {
+  override def getInfoDateColumn(df: DataFrame): Column = current_timestamp().cast(DateType)
+}
+
+object InfoDateFactory {
+  private val defaultEventTimestampPattern = "yyyy-MM-dd'T'HH:mm'Z'"
+
+  val log: Logger = LoggerFactory.getLogger(this.getClass)
+
+  def getFactoryFromConfig(conf: Configuration): InfoDateFactory = {
+    if (conf.containsKey(reportDateKey)) {
+      val reportDate = conf.getString(reportDateKey)
+      log.info(s"Information date: Explicit from the job configuration = $reportDate")
+      new InfoDateLiteralFactory(reportDate)
+    } else if (conf.containsKey(eventTimestampColumnKey)) {
+      val eventTimestampColumnName = conf.getString(eventTimestampColumnKey)
+      val eventTimestampPattern = conf.getString(eventTimestampPatternKey, defaultEventTimestampPattern)
```

Comment:
I wonder if it is worth to somehow validate the supplied pattern beforehand in order to detect possibly invalid TS pattern before spark processing is started (aim: to fail fast).

Reply:
I'm open to suggestions

Reply:
Is it even possible to validate patterns? I think it's like email validation. Too much time, too much effort, low returns.

Reply:
Well, since `to_timestamp` uses the pattern rules of `java.text.SimpleDateFormat` which does not provide a public method to check pattern validity, the easiest way would be to construct a `new SimpleDateFormat(pattern)` to see if it fails with an exception.

The more I think about it though, the less I am convinced that it is worth it. There is not a lot that can fail while the pattern is read (non-date chars are treated as normal chars anyway) and the key part where it would most probably fail (the `to_timestamp` application) cannot be known beforehand without the data.

All in all, failing fast is hard to do here. :(